### PR TITLE
Optimize multi stage docker build

### PR DESCRIPTION
**Pull request by KevinMind**
Relates to: https://mozilla-hub.atlassian.net/browse/ADDSRV-720
Partially fulfills: https://github.com/mozilla/addons-server/pull/22182

### Description

This pull request includes changes to optimize the multi stage docker build process. It improves efficiency and reduces resource usage. It improves readability of the dockerfile and adds additional checks to the built image to verify that the expected files/packages are installed in the final image.

1) separate assets stage (better caching)
2) final stage copies from other stages
3) move host copying to latest possible stage
4) add make commands to verify image state is correct
5) re-organize dockerfile for readability

### Context

This PR not only improves the build currently but paves the way for defining a development/production final stage split allowing us to efficiently build containers for development or production purposes without invalidating the majority of the build.

### Testing

Test by running ci or building locally, inspecting the files and commands.

1. Smoke test, can you build the image and run the container
2. Tests, are they passing
3. Are files where they are supposed to be

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -64,29 +64,29 @@ ENV LC_ALL en_US.UTF-8
...
-RUN mkdir /deps \
-    && chown -R olympia:olympia /deps \
-    && rm -rf ${HOME}/src/olympia.egg-info \
-    && mkdir -p ${HOME}/src/olympia.egg-info \
-    && chown olympia:olympia ${HOME}/src/olympia.egg-info \
-    # For backwards-compatibility purposes, set up links to uwsgi. Note that
-    # the target doesn't exist yet at this point, but it will later.
-    && ln -s /deps/bin/uwsgi /usr/bin/uwsgi \
-    && ln -s /usr/bin/uwsgi /usr/sbin/uwsgi
+RUN <<EOF
```
Comment (KevinMind) [inline Dockerfile:73]: Just a cleaner way of writing the same code. + adding the symlink for package*.json

File: Dockerfile [hunk 2]
```diff
@@ -99,23 +99,30 @@ ENV NPM_DEBUG=true
...
     # Files needed to run the make command
-    --mount=type=bind,source=Makefile,target=${HOME}/Makefile \
     --mount=type=bind,source=Makefile-docker,target=${HOME}/Makefile-docker \
     # Files required to install pip dependencies
-    --mount=type=bind,source=setup.py,target=${HOME}/setup.py \
-    --mount=type=bind,source=./requirements,target=${HOME}/requirements \
+    --mount=type=bind,source=./requirements/pip.txt,target=${HOME}/requirements/pip.txt \
+    --mount=type=cache,target=${PIP_CACHE_DIR},uid=${OLYMPIA_UID},gid=${OLYMPIA_UID} \
+    # Command to install dependencies
+    make -f Makefile-docker update_deps_pip
```
Comment (KevinMind) [inline Dockerfile:107]: All we need in "base" is pip to be installed, this let's other layers install packages using the correct version.

File: Dockerfile [hunk 3]
```diff
@@ -99,23 +99,30 @@ ENV NPM_DEBUG=true
...
+
+FROM base as pip_production
```
Comment (KevinMind) [inline Dockerfile:109]: Define production dependencies as a single layer, let's the rest of the stages inherit prod dependencies and makes copying the /deps dir to the final layer easy.

File: Dockerfile [hunk 4]
```diff
@@ -2,91 +2,75 @@
...
+rm -rf /var/lib/apt/lists/*
 
-# Add our custom mime types (required for for ts/json/md files)
-ADD docker/etc/mime.types /etc/mime.types
+export APT_KEY_DONT_WARN_ON_DANGEROUS_USAGE=DontWarn
+apt-key add /etc/pki/gpg/nodesource.gpg.key
+apt-key add /etc/pki/gpg/mysql.gpg.key
+EOF
+
+RUN --mount=type=bind,source=docker/dependencies.txt,target=/dependencies.txt <<EOF
```
[the remark this answers is not in the document]
Comment (KevinMind) [reply Dockerfile:38]: I would hope in a future iteration we actually split the file into prod and dev and maybe even build. But agree on the name.

I think splitting the files with reasonable names could get you part way there in terms of losing comments.

There's also a way I can support comments in that file (just need to ignore lines that start with # so we could do that as well..

wdyt? 
1. reanem to debian_pacakges.txt
2. support comments
3. re-add the comments

Is that a plan for now?

[the remark this answers is not in the document]
Comment (diox) [reply Dockerfile:38]: Yes renaming and supporting basic comments (re-adding those we already had) would be enough for now

File: Dockerfile [hunk 5]
```diff
@@ -64,29 +64,29 @@ ENV LC_ALL en_US.UTF-8
...
+mkdir /deps
+chown -R olympia:olympia /deps
+rm -rf ${HOME}/src/olympia.egg-info
+mkdir -p ${HOME}/src/olympia.egg-info
+chown olympia:olympia ${HOME}/src/olympia.egg-info
+# For backwards-compatibility purposes, set up links to uwsgi. Note that
+# the target doesn't exist yet at this point, but it will later.
+ln -s /deps/bin/uwsgi /usr/bin/uwsgi
+ln -s /usr/bin/uwsgi /usr/sbin/uwsgi
+# add symlink for old uwsgi binary paths
```
[the remark this answers is not in the document]
Comment (KevinMind) [reply Dockerfile:67]: Done.

File: Dockerfile [hunk 6]
```diff
@@ -2,91 +2,75 @@
 # Read the docs/topics/development/docker.md file for more information about this Dockerfile.
 ####################################################################################################
 
-FROM python:3.10-slim-bookworm as base
+FROM python:3.10-slim-bookworm as olympia
 
 # Should change it to use ARG instead of ENV for OLYMPIA_UID/OLYMPIA_GID
 # once the jenkins server is upgraded to support docker >= v1.9.0
```
[the remark this answers is not in the document]
Comment (KevinMind) [reply Dockerfile:8]: Done.

File: Dockerfile [hunk 7]
```diff
@@ -2,91 +2,75 @@
...
+
+# give olympia access to the HOME directory
+ENV HOME /data/olympia
+WORKDIR ${HOME}
+RUN chown -R olympia:olympia ${HOME}
+
+FROM olympia as base
 # Add keys and repos for node and mysql
+# TODO: replace this with a bind mount on the RUN command
 COPY docker/*.gpg.key /etc/pki/gpg/
```
[the remark this answers is not in the document]
Comment (KevinMind) [reply Dockerfile:24]: Done.

File: scripts/check_version_json.sh [hunk 8]
```diff
@@ -0,0 +1,19 @@
...
+
+REQUIRED_VERSION_KEYS=(
+  version
+  build
+  commit
+  source
+)
+
+for key in "${REQUIRED_VERSION_KEYS[@]}"; do
+  if ! grep -q "\"$key\":" version.json; then
```
[the remark this answers is not in the document]
Comment (KevinMind) [reply scripts/check_version_json.sh]: Done

File: Dockerfile (hunk 4 above)
[the remark this answers is not in the document]
Comment (KevinMind) [reply Dockerfile]: Done.

File: Dockerfile (hunk 7 above)
[the remark this answers is not in the document]
Comment (KevinMind) [reply Dockerfile]: Can you verify that the changes applied are what you were looking for?

**Comment by diox** [comment]
Mmmm it seems I somehow managed to have a polluted cache, I'm running into uwsgi build issues again:

```
% make build_docker_image
docker buildx use container 2>/dev/null || docker buildx create \
    --name container \
    --driver=docker-container
./scripts/version.sh local 9cb84999e7a87a763e76eb7af7c083073191e7a3 build
+ version=local
+ commit=9cb84999e7a87a763e76eb7af7c083073191e7a3
+ build=build
+ cat
+ cat version.json
{
  "commit": "9cb84999e7a87a763e76eb7af7c083073191e7a3",
  "version": "local",
  "build": "build",
  "source": "https://github.com/mozilla/addons-server"
}
DOCKER_BUILDKIT=1 docker buildx build -t addons-server-test --load --platform linux/amd64 --progress=auto --builder=container --label git.commit=9cb84999e7a87a763e76eb7af7c083073191e7a3 --cache-to=type=local,dest=docker-cache-new  --cache-from=type=local,src=docker-cache,mode=max .
[+] Building 0.0s (0/0)      docker-container:container
[+] Building 133.8s (35/35) FINISHED                                                                                                                                                                                               docker-container:container
 => [internal] load build definition from Dockerfile                                                                                                                                                                                                     0.0s
 => => transferring dockerfile: 6.59kB                                                                                                                                                                                                                   0.0s
 => [internal] load metadata for docker.io/library/python:3.10-slim-bookworm                                                                                                                                                                             0.8s
 => [auth] library/python:pull token for registry-1.docker.io                                                                                                                                                                                            0.0s
 => [internal] load .dockerignore                                                                                                                                                                                                                        0.0s
 => => transferring context: 216B                                                                                                                                                                                                                        0.0s
 => importing cache manifest from local:14183617221189516797                                                                                                                                                                                             0.0s
 => => inferred cache manifest type: application/vnd.oci.image.index.v1+json                                                                                                                                                                             0.0s
 => CACHED [internal] settings cache mount permissions                                                                                                                                                                                                   0.0s
 => [internal] load build context                                                                                                                                                                                                                        2.3s
 => => transferring context: 27.07MB                                                                                                                                                                                                                     2.1s
 => [olympia 1/4] FROM docker.io/library/python:3.10-slim-bookworm@sha256:161d1e3deddba156d7834d15e6fe4d010b3b6f69f18ee1cfba9348e712f4a4ba                                                                                                               0.0s
 => => resolve docker.io/library/python:3.10-slim-bookworm@sha256:161d1e3deddba156d7834d15e6fe4d010b3b6f69f18ee1cfba9348e712f4a4ba                                                                                                                       0.0s
 => CACHED [olympia 2/4] RUN <<EOF (groupadd -g 9500 olympia...)                                                                                                                                                                                         0.0s
 => CACHED [olympia 3/4] WORKDIR /data/olympia                                                                                                                                                                                                           0.0s
 => CACHED [olympia 4/4] RUN chown -R olympia:olympia /data/olympia                                                                                                                                                                                      0.0s
 => CACHED [base 1/7] COPY docker/*.gpg.asc /etc/apt/trusted.gpg.d/                                                                                                                                                                                      0.0s
 => CACHED [base 2/7] COPY docker/*.list /etc/apt/sources.list.d/                                                                                                                                                                                        0.0s
 => [base 3/7] RUN <<EOF (# Add support for https apt repos and gpg signed repos...)                                                                                                                                                                     5.2s
 => [base 4/7] RUN --mount=type=bind,source=docker/debian_packages.txt,target=/debian_packages.txt /bin/bash  <<EOF                                                                                                                                     45.8s 
 => [base 5/7] RUN localedef -i en_US -f UTF-8 en_US.UTF-8                                                                                                                                                                                               1.1s 
 => [base 6/7] RUN <<EOF (# Create directory for dependencies...)                                                                                                                                                                                        0.1s 
 => [base 7/7] RUN     --mount=type=bind,source=Makefile-docker,target=/data/olympia/Makefile-docker     --mount=type=bind,source=./requirements/pip.txt,target=/data/olympia/requirements/pip.txt     --mount=type=cache,target=/deps/cache/,uid=9500,  1.9s 
 => [sources 1/6] RUN --mount=type=bind,src=scripts/generate_build.py,target=/generate_build.py     /generate_build.py > build.py                                                                                                                        0.2s 
 => [pip_production 1/1] RUN     --mount=type=bind,source=Makefile-docker,target=/data/olympia/Makefile-docker     --mount=type=bind,source=./requirements/prod.txt,target=/data/olympia/requirements/prod.txt     --mount=type=bind,source=package.js  17.7s 
 => [sources 2/6] COPY docker/etc/mime.types /etc/mime.types                                                                                                                                                                                             0.1s 
 => [sources 3/6] COPY --chown=olympia:olympia . /data/olympia                                                                                                                                                                                           7.2s 
 => [assets 1/3] RUN     --mount=type=bind,source=Makefile-docker,target=/data/olympia/Makefile-docker     --mount=type=bind,source=setup.py,target=/data/olympia/setup.py     --mount=type=bind,source=pyproject.toml,target=/data/olympia/pyproject.t  2.2s 
 => [locales 1/2] COPY --chown=olympia:olympia locale /data/olympia/locale                                                                                                                                                                               0.4s 
 => [locales 2/2] RUN     --mount=type=bind,source=Makefile-docker,target=/data/olympia/Makefile-docker     --mount=type=bind,source=locale/compile-mo.sh,target=/data/olympia/compile-mo.sh     --mount=type=bind,source=requirements/locale.txt,targ  10.9s 
 => [assets 2/3] COPY --chown=olympia:olympia static/ /data/olympia/static/                                                                                                                                                                              0.3s 
 => [assets 3/3] RUN     --mount=type=bind,src=src,target=/data/olympia/src     --mount=type=bind,src=Makefile-docker,target=/data/olympia/Makefile-docker     --mount=type=bind,src=manage.py,target=/data/olympia/manage.py     echo "from olympia.li  9.3s 
 => [sources 4/6] COPY --from=locales --chown=olympia:olympia /data/olympia/locale /data/olympia/locale                                                                                                                                                  0.5s 
 => [sources 5/6] COPY --from=assets --chown=olympia:olympia /data/olympia/site-static /data/olympia/site-static                                                                                                                                         0.5s 
 => [sources 6/6] COPY version.json /app/version.json                                                                                                                                                                                                    0.1s 
 => [production 1/2] COPY --from=pip_production --chown=olympia:olympia /deps /deps                                                                                                                                                                      3.7s 
 => [production 2/2] RUN make -f Makefile-docker update_deps_olympia                                                                                                                                                                                     2.2s 
 => exporting to docker image format                                                                                                                                                                                                                    36.1s 
 => => exporting layers                                                                                                                                                                                                                                 19.3s 
 => => exporting manifest sha256:7f2e50f244cda814fb85cd219f65d66f1e2a41b38780ce01a4744f9613b83feb                                                                                                                                                        0.0s 
 => => exporting config sha256:2f2a64596af9d4aad4302be9efd512bab1519e3d1f0cbee4c88817ae211ae7d9                                                                                                                                                          0.0s 
 => => sending tarball                                                                                                                                                                                                                                  16.8s 
 => importing to docker                                                                                                                                                                                                                                 15.0s 
 => => loading layer 85d60abc0ee4 65.54kB / 3.43MB                                                                                                                                                                                                      15.0s 
 => => loading layer e9d293ea8ce6 168.23MB / 283.75MB                                                                                                                                                                                                   14.8s
 => => loading layer d8780a75b06b 32.77kB / 863.24kB                                                                                                                                                                                                    11.0s
 => => loading layer 8096caad693b 335B / 335B                                                                                                                                                                                                           10.9s
 => => loading layer a17e7118ea64 65.54kB / 3.77MB                                                                                                                                                                                                      10.9s
 => => loading layer 4e1a520f8a54 215B / 215B                                                                                                                                                                                                           10.7s
 => => loading layer 3ef5d8f7504c 7.19kB / 7.19kB                                                                                                                                                                                                       10.7s
 => => loading layer a0d3d992a9e9 215.02MB / 317.86MB                                                                                                                                                                                                   10.6s
 => => loading layer c96ed1c65383 65.54kB / 3.51MB                                                                                                                                                                                                       5.0s
 => => loading layer 663472752769 557.06kB / 55.96MB                                                                                                                                                                                                     4.9s
 => => loading layer 6b3490a2ab24 254B / 254B                                                                                                                                                                                                            4.4s
 => => loading layer 4701ea545346 59.60MB / 90.84MB                                                                                                                                                                                                      4.4s
 => => loading layer 1680674dfc24 32.77kB / 45.14kB                                                                                                                                                                                                      0.2s
 => exporting cache to client directory                                                                                                                                                                                                                  3.7s
 => => preparing build cache for export                                                                                                                                                                                                                  3.7s
 => => writing layer sha256:0a654df49bbd4a4149d0333e744d68f7abbeecdba1920804c2eec39e938d0818                                                                                                                                                             0.3s
 => => writing layer sha256:178c83b1d90a5b3235e0a86a131ea5165d02874d373a93c1fbd6353d4bbedf06                                                                                                                                                             0.0s
 => => writing layer sha256:2556d273785e2fb46fdf235ed74a2b70f33d5f7665e49e8b9f6c9fbfb6b312cd                                                                                                                                                             0.4s
 => => writing layer sha256:2e97c7a1dca91a3679ee997d2a8f547ba77089a88bb09451b2e4566dd3498010                                                                                                                                                             0.0s
 => => writing layer sha256:4f1781d93e6842b8234414f363df28625d770c750fa1a52eb9631b971a5c0fe6                                                                                                                                                             0.0s
 => => writing layer sha256:72914424168c8ebb0dbb3d0e08eb1d3b5b2a64cc51745bd65caf29c335b31dc7                                                                                                                                                             0.0s
 => => writing layer sha256:78d35c3f33512291b7b4dd47378f5911a6f6de25d2dbd251d5bcf3314ea39db0                                                                                                                                                             0.0s
 => => writing layer sha256:7a9a6f0ed9c9a9c9439fcfbc8d8fe9c904612843059776e9fa33e4e3843f1d8e                                                                                                                                                             0.0s
 => => writing layer sha256:7d52992ba7c286753f27a1c36d6d49c0f338c6c21c53cc5f86c60cb5a30f30c5                                                                                                                                                             0.0s
 => => writing layer sha256:8007f21c2e09cdcc8c027b08004e0c49fd29eddf7ff6513a74adf07540247b32                                                                                                                                                             0.0s
 => => writing layer sha256:878ad1a70557dd8bf66c2b7f30fc1205586f0f29f0216547285f241d7352d402                                                                                                                                                             0.0s
 => => writing layer sha256:90d858cabd4a0f50be68d7c0a1e4ca46bc862d9f172deb47a44a255d1ddd4b02                                                                                                                                                             1.3s
 => => writing layer sha256:b03ab4285c5b0b55b102670ef70c7f3d5ea65d18c8ba15dd51ce2574e9b1513c                                                                                                                                                             0.0s
 => => writing layer sha256:b0a0cf830b12453b7e15359a804215a7bcccd3788e2bcecff2a03af64bbd4df7                                                                                                                                                             0.0s
 => => writing layer sha256:b5b17c82b9f364b619b5d052d10764f374f648fc9aedc0027d13f600432d6dd4                                                                                                                                                             0.0s
 => => writing layer sha256:b9154e4b652b573a29324bbad86414cbe6e674d3a34b91a43808f7cae9994395                                                                                                                                                             0.0s
 => => writing layer sha256:dfa8d44b86cc57dcb19419bfe4868e4958dc36dac2923aa0d38e91e4ca9a9d87                                                                                                                                                             1.2s
 => => writing layer sha256:e2905cc278f1c6d635140e5aec2698a1d9ebd0e4b45314b3b08bb4f32ecadc75                                                                                                                                                             0.0s
 => => writing layer sha256:ed89fd7c2cef21bd8f5eb784a81d1531421a345d75514e1e468eed3ac1df1dad                                                                                                                                                             0.0s
 => => writing layer sha256:ee97fdf7df389f77b1e024b886115233a20247ca968c272ef1c668cf635d35fc                                                                                                                                                             0.0s
 => => writing layer sha256:f1233a216289c28a81d7a45a28b0553260b9c19f762bb4fcd02ba2f13b444d50                                                                                                                                                             0.0s
 => => writing layer sha256:f3856efbcea28ab9a3c266a87b504bb81cf460dd985fe46054f993ac80be8ebb                                                                                                                                                             0.0s
 => => writing layer sha256:f38a48c634e28ec3277853c8dd06d9d5c26745a4ca57ea946063d04ec4adef61                                                                                                                                                             0.0s
 => => writing config sha256:ca5fa363e85c0c4f40b1069e11778524e349af8cdc64b4145752b2c18e315ed1                                                                                                                                                            0.0s
 => => writing cache manifest sha256:7fe92e98a0b8b3c03a6e8ddd30ae885e09785e44c935c37f7c432e90453aafe9                                                                                                                                                    0.0s
rm -rf docker-cache
mv docker-cache-new docker-cache

% docker compose stop
(...)

% docker image tag addons-server-test mozilla/addons-server:latest   

% docker compose up -d
(...)

% docker compose logs web
(...)
web-1  | Updating olympia user to host defined UID/GID...
web-1  | UID: 1000
web-1  | GID: 1000
web-1  | 2024-05-06 09:18:09,162 INFO Set uid to user 1000 succeeded
web-1  | 2024-05-06 09:18:09,172 INFO RPC interface 'supervisor' initialized
web-1  | 2024-05-06 09:18:09,172 CRIT Server 'inet_http_server' running without any HTTP authentication checking
web-1  | 2024-05-06 09:18:09,173 INFO supervisord started with pid 21
web-1  | 2024-05-06 09:18:10,175 INFO spawned: 'olympia' with pid 22
web-1  | 2024-05-06 09:18:10,177 INFO spawned: 'watcher' with pid 23
web-1  | 2024-05-06 09:18:10,181 WARN exited: olympia (exit status 127; not expected)
web-1  | 2024-05-06 09:18:11,184 INFO spawned: 'olympia' with pid 27
web-1  | 2024-05-06 09:18:11,184 INFO success: watcher entered RUNNING state, process has stayed up for > than 1 seconds (startsecs)
web-1  | 2024-05-06 09:18:11,188 WARN exited: olympia (exit status 127; not expected)
web-1  | 2024-05-06 09:18:13,192 INFO spawned: 'olympia' with pid 28
web-1  | 2024-05-06 09:18:13,195 WARN exited: olympia (exit status 127; not expected)
web-1  | 2024-05-06 09:18:16,199 INFO spawned: 'olympia' with pid 29
web-1  | 2024-05-06 09:18:16,202 WARN exited: olympia (exit status 127; not expected)
web-1  | 2024-05-06 09:18:16,202 INFO gave up: olympia entered FATAL state, too many start retries too quickly

% tail logs/docker-olympia.log 
(...)
uwsgi: error while loading shared libraries: libssl.so.1.1: cannot open shared object file: No such file or directory
 % docker compose exec web ./manage.py check
(...)
SystemCheckError: System check identified some issues:

ERRORS:
?: (setup.E001) uwsgi --version returned a non-zero value

System check identified 1 issue (21 silenced).
```

**Comment by diox** [comment]
For posterity: ran into an issue with some deps because I had no cleaned up `deps/` directory before rebuilding. Ended up with a broken `uwsgi` that had been built from an old image...

**Review by diox — APPROVED** [review]

**Pull request merged**